File: ch7/7-7.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const (
	MaxLine    = 1000
	PatternLen = 100
)

func main() {
	var f *os.File
	var lineno int
	var c rune
	var except, number, pat_found bool
	var pattern, filename string
	var err error

	for i := 1; i < len(os.Args); i++ {
		if os.Args[i][0] == '-' {
			c = rune(os.Args[i][1])
			switch c {
			case 'x':
				except = true
				break
			case 'n':
				number = true
				break
			default:
				fmt.Printf("find: illegal option %c\n", c)
				os.Exit(1)
				break
			}
		} else {
			if !pat_found {
				pattern = os.Args[i]
				pat_found = true
			} else {
				filename = os.Args[i]
			}
		}
	}
	f, err = os.Open(filename)
	if err != nil {
		fmt.Printf("find: %s\n", err)
		os.Exit(1)
	}
	s := bufio.NewScanner(f)
	for s.Scan() {
		l := s.Text()
		if except {
			// do something
		}
		if strings.Contains(l, pattern) {
			if number {
				lineno++
				fmt.Printf("%d:", lineno)
			}
			fmt.Printf("%s\n", l)
		}
	}
}

func getLine(lim int) ([]byte, int) {
	var c byte
	var i int
	s := make([]byte, lim)
	r := bufio.NewReader(os.Stdin)
	for i = 0; i < lim-1; i++ {
		c, err := r.ReadByte()
		if rune(c) == '\n' || err != nil {
			break
		}
		s[i] = c
	}
	if rune(c) == '\n' {
		s[i] = c
		i++
	}
	return s, i
}

File: kr/ch7/7-7.c
#include <stdio.h>
#include <string.h>

#define MAXLINE 1000
#define PATTERNLEN 100

int get_line(char *line, int max);

int main(int argc, char *argv[])
{
    FILE *f;
    char line[MAXLINE], *pattern;
    long lineno = 0;
    int c, except = 0, number = 0, found = 0;
    int argn = 1;
    while (++argn < argc && ((*++argv)[0] == '-')) {
        while ((c = *++argv[0])) {
            switch (c) {
                case 'x':
                    except = 1;
                    break;
                case 'n':
                    number = 1;
                    break;
                default:
                    printf("find: illegal option %c\n", c);
                    argc = 0;
                    found = -1;
                    break;
            }
        }
    }

    if (argn++ <= argc) {
        pattern = *argv;
        if (argn <= argc) {
            lineno = 0;
            while (*++argv != NULL) {
                printf("%s\n", *argv);
                if ((f = fopen(*argv, "r")) != NULL) {
                    while (fgets(line, MAXLINE, f) != NULL) {
                        lineno++;
                        if ((strstr(line, pattern) != NULL) != except) {
                            if (number) {
                                printf("%ld:", lineno);
                            }
                            printf("%s", line);
                            found++;
                        }
                    }
                } else {
                    printf("error: could not open %s", * argv);
                }
            }
        } else {
            while (get_line(line, MAXLINE) > 0) {
                lineno++;
                if ((strstr(line, pattern) != NULL) != except) {
                    if (number) {
                        printf("%ld:", lineno);
                    }
                    printf("%s", line);
                    found++;
                }
            }
        }
    } else {
        printf("Usage: find -x -n pattern [textfile\\s]\n");
    }
    return found;
}

int get_line(char * s, int lim)
{
    int c, i;
    i = 0;
    while (--lim > 0 && (c = getchar()) != EOF && c != '\n') {
        s[i++] = c;
    }
    if (c == '\n') {
        s[i++] = c;
    }
    s[i] = '\0';
    return i;
}
